crates/red_knot_workspace/src/site_packages.rs

```diff
@@ -0,0 +1,130 @@
+//! Utilities for finding the `site-packages` directory,
+//! into which third-party packages are installed.
+//!
+//! The routines exposed by this module have different behaviour depending
+//! on the platform of the *host machine*, which may be
+//! different from the *target platform for type checking*. (A user
+//! might be running red-knot on a Windows machine, but might
+//! reasonably ask us to type-check code assuming that the code runs
+//! on Linux.)
+
+use std::io;
+
+use ruff_db::system::{System, SystemPath, SystemPathBuf};
+
+/// Attempt to retrieve the `site-packages` directory that is actually
+/// inside the virtual environment.
+///
+/// It is assumed that it has already been checked that `venv_path` points
+/// to an existing directory on disk.
+#[cfg(target_os = "windows")]
+fn venv_site_packages_dir(
```

Comment:
This routine (at least as currently written) isn't really specific to a venv, it's just encoding knowledge about Python installation layouts in general. Given a Python binary, you go up one directory to find the "base" of the installation (which could be a venv, or not), and there you expect to find `lib/` or `Lib/`. I would be inclined to name the functions accordingly and not over-fit to the case of a virtualenv; we should also support non-virtualenv Pythons.

Reply:
Are you sure you can generalise to that extent? For example, here's what I get with my Brew-installed system Python installation:

```pycon
/dev % /opt/homebrew/bin/python
Python 3.12.4 (main, Jun  6 2024, 18:26:44) [Clang 15.0.0 (clang-1500.3.9.4)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import site, sys
>>> sys.executable
'/opt/homebrew/opt/python@3.12/bin/python3.12'
>>> site.getsitepackages()
['/opt/homebrew/opt/python@3.12/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages']
```

Reply:
Pretty sure, yeah :) The behavior of "start with binary location, go up one directory, then look for `lib/...` (specifically it looks for the `os.py` file in the standard library) is hardcoded directly into CPython as the way to determine where the standard library is found in the first place, so someone would have to change that code in CPython itself to change that behavior.

Of course it is possible to modify `site.py` (or do whatever you want in `sitecustomize.py`) to divorce the location of site-packages from the `lib` directory, but that kind of trickery we can't handle at all with the approach in this PR, the only way to handle it is to run Python and query it, like `uv` does.

In your Homebrew case, I suspect you'll find that `/opt/homebrew/opt/python@3.12/bin/python3.12` is a symlink (possibly via several symlink steps) to `/opt/homebrew/opt/python@3.12/Frameworks/Python.framework/Versions/3.12/bin/python`. You do have to resolve symlinks of the python binary to get this correct in general, because CPython does.

Reply:
> Of course it is possible to modify `site.py` (or do whatever you want in `sitecustomize.py`) to divorce the location of site-packages from the `lib` directory, but that kind of trickery we can't handle at all with the approach in this PR, the only way to handle it is to run Python and query it, like `uv` does.

Right, but I understood that this kind of "trickery" was pretty common for system Python installations provided by Homebrew and Linux distros. Common enough, I'd argue, that you can only guarantee that this routine -- as it currently stands -- will work for a virtual environment, not really for an arbitrary system installation. I think the "standard" way Python installation layouts are "meant" to be is pretty irrelevant if that doesn't reflect the experience of people who actually have system installations of Python.

Reply:
I don't think it is common at all to move site-packages elsewhere; I've never heard of a distro (or Homebrew) doing that. Debian adds an extra `dist-packages` directory; that's the only similar case I've heard of; we can special-case that. (I'm not sure if you processed the last paragraph of my response, which clarifies that your case is not an example of Homebrew breaking the usual layout; it's following the usual layout, once you follow the symlinks on the binary.)

The more important point is, however we handle system Python oddities, we have to handle them regardless, if we are to support `include-system-site-packages` in venvs. So if we have to make at least a best effort to handle system Pythons regardless, why treat venvs specially?

Reply:
> In your Homebrew case, I suspect you'll find that `/opt/homebrew/opt/python@3.12/bin/python3.12` is a symlink (possibly via several symlink steps) to `/opt/homebrew/opt/python@3.12/Frameworks/Python.framework/Versions/3.12/bin/python`. You do have to resolve symlinks of the python binary to get this correct in general, because CPython does.

Looks like there are some symlinks in there, yup, but even after resolving the symlinks, the Python executable is nowhere near the `site-packages` directory:

```pycon
/dev % /opt/homebrew/bin/python
Python 3.12.4 (main, Jun  6 2024, 18:26:44) [Clang 15.0.0 (clang-1500.3.9.4)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import pathlib, sys, site
>>> pathlib.Path(sys.executable).resolve()
PosixPath('/opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/bin/python3.12')
>>> pathlib.Path(site.getsitepackages()[0]).resolve()
PosixPath('/opt/homebrew/lib/python3.12/site-packages')
```

Reply:
What is in `/opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/`?

And note that `/opt/homebrew/lib/python3.12/site-packages` is exactly where you'd expect to find it relative to `/opt/homebrew/bin/python`.

I don't have access to your system, but I am pretty sure that while Homebrew may do some symlinking stuff, it doesn't hack site-packages location, it sets it up in a way that works with CPython's lib-finding code.

But like I said above, the more important point is that we have to solve these complexities either way, if we will support include-system-site-packages. So we can't get out of handling these by claiming to only support virtualenvs.

Reply:
> But like I said above, the more important point is that we have to solve these complexities either way, if we will support include-system-site-packages. So we can't get out of handling these by claiming to only support virtualenvs.

`include-system-site-packages` is pretty rare. I was wondering about simply invoking Python to figure out where the system `site-packages` are if we find that setting in the `pyvenv.cfg` file. But I suppose we could try the static method first, and then if we can't find a directory where we "expect", just fall back to invoking Python.

Reply:
> What is in `/opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib`?

Looks like we've got a `libpython3.12.dylib` file, a `pkgconfig` directory with some `.pc` files in it, and a `python3.12` directory that contains the standard library

Reply:
> and a python3.12 directory that contains the standard library

Within which (if your Homebrew install is like mine) there's a `site-packages` which is a symlink all the way back to `/opt/homebrew/lib/python3.12/site-packages` :) So yeah, Homebrew's just playing a lot of symlinking tricks in a way that still works as a normal layout.

> I was wondering about simply invoking Python to figure out where the system site-packages are if we find that setting in the pyvenv.cfg file. But I suppose we could try the static method first, and then if we can't find a directory where we "expect", just fall back to invoking Python.

Ah, makes sense. I do think it's reasonable if you prefer to only use the static-finding approach for virtualenvs, and always fall back to invoking Python for system Pythons. I agree include-system-site-packages isn't super common, and venvs should be the most common scenario we deal with.

Reply:
> Within which (if your Homebrew install is like mine) there's a `site-packages` which is a symlink all the way back to `/opt/homebrew/lib/python3.12/site-packages` :) So yeah, Homebrew's just playing a lot of symlinking tricks in a way that still works as a normal layout.

I see... you're quite right!

I think I'll rename this routine to something more generic (so that we can reuse it for system site-packages, as you suggest), but keep the CLI command as `--venv-path` (unless you have suggestions in response to my question at https://github.com/astral-sh/ruff/pull/12716#discussion_r1705919180)